### PR TITLE
Adaptive read task size calculation in parallel replicas

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Use adaptive read task size calculation method (adaptive meaning it depends on read column sizes) for parallel replicas.

---

We use sizes of columns from query to determine task size in both `MergeTreeReadPool` and `MergeTreePrefetchedReadPool` for remote reading. There doesn't seem to be any reason not to do it for parallel replicas too.
This pr moves this logic into `MergeTreeReadPoolBase` and then supports it in `MergeTreeReadPoolParallelReplicas*`.

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 5fc648a07cfcf6229f5cf6ce8dc2cd6ba2d090c8 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60377/5fc648a07cfcf6229f5cf6ce8dc2cd6ba2d090c8/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60377/5fc648a07cfcf6229f5cf6ce8dc2cd6ba2d090c8/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Review by alesapin — APPROVED** [review]

**Comment by nickitat** [comment]
`AST fuzzer (tsan)` - some random abort of ch client
`Stateless tests (tsan) [1/5]` - #60625

**Comment by alexey-milovidov** [comment]
@nickitat, this is a performance improvement, not "not for changelog". Please edit it accordingly.

File: src/Storages/MergeTree/MergeTreeReadPoolBase.cpp [hunk 1]
```diff
@@ -34,10 +38,54 @@ MergeTreeReadPoolBase::MergeTreeReadPoolBase(
...
+static size_t getApproxSizeOfPart(const IMergeTreeDataPart & part, const Names & columns_to_read)
+{
+    ColumnSize columns_size{};
+    for (const auto & col_name : columns_to_read)
+        columns_size.add(part.getColumnSize(col_name));
+    /// For compact parts we don't know individual column sizes, let's use whole part size as approximation
+    return columns_size.data_compressed ? columns_size.data_compressed : part.getBytesOnDisk();
+}
+
+static size_t calculateMinMarksPerTask(
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeReadPoolBase.cpp:53]: I would name it "adjustMinMarksPerTask", but up to you

File: src/Storages/MergeTree/MergeTreeReadPoolBase.cpp [hunk 2]
```diff
@@ -34,10 +38,54 @@ MergeTreeReadPoolBase::MergeTreeReadPoolBase(
...
+    const Names & columns_to_read,
+    PrewhereInfoPtr prewhere_info,
+    const MergeTreeReadPoolBase::PoolSettings & pool_settings,
+    const Settings & settings)
+{
+    size_t min_marks_per_task = pool_settings.min_marks_for_concurrent_read;
+    const size_t part_marks_count = part.getMarksCount();
+    if (part_marks_count && part.data_part->isStoredOnRemoteDisk())
+    {
+        const auto & columns = settings.merge_tree_determine_task_size_by_prewhere_columns && prewhere_info
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeReadPoolBase.cpp:64]: Let's add a comment with idea explanation.

File: src/Storages/MergeTree/MergeTreeReadPoolBase.cpp [hunk 3]
```diff
@@ -34,10 +38,54 @@ MergeTreeReadPoolBase::MergeTreeReadPoolBase(
...
+            ? prewhere_info->prewhere_actions->getRequiredColumnsNames()
+            : columns_to_read;
+        const size_t part_compressed_bytes = getApproxSizeOfPart(*part.data_part, columns);
+
+        const auto avg_mark_bytes = std::max<size_t>(part_compressed_bytes / part_marks_count, 1);
+        const auto min_bytes_per_task = settings.merge_tree_min_bytes_per_task_for_remote_reading;
+        /// We're taking min here because number of tasks shouldn't be too low - it will make task stealing impossible.
+        const auto heuristic_min_marks
+            = std::min<size_t>(pool_settings.sum_marks / pool_settings.threads / 2, min_bytes_per_task / avg_mark_bytes);
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeReadPoolBase.cpp:73]: Why `/ 2`?

**Review by alesapin — APPROVED** [review]

File: src/Core/SettingsChangesHistory.cpp [hunk 4]
```diff
@@ -76,6 +76,7 @@ static std::initializer_list<std::pair<ClickHouseVersion, SettingsChangesHistory
               {"azure_sdk_max_retries", 10, 10, "Maximum number of retries in azure sdk"},
               {"azure_sdk_retry_initial_backoff_ms", 10, 10, "Minimal backoff between retries in azure sdk"},
               {"azure_sdk_retry_max_backoff_ms", 1000, 1000, "Maximal backoff between retries in azure sdk"},
+              {"merge_tree_min_bytes_per_task_for_remote_reading", 4194304, 2097152, "Value is unified with `filesystem_prefetch_min_bytes_for_single_read_task`"},
```
Comment (Algunenano) [inline src/Core/SettingsChangesHistory.cpp:79]: This went into 24.8, not 24.7

**Pull request merged**